### ISSUE TITLE
Samsung is NOT worst

### ISSUE THREAD

**Issue by T1mmyK0** [urbandroid-team/dont-kill-my-app#804]
Hey,

I have a big problem. I am linking to dontkillmyapp.com from my Android app.
My app has over 5 million downloads, so I have some experience.
I have noticed that actually Samsung devices work best, they have the least problems and I also recommend them whenever someone asks. I am also using a Samsung Galaxy S22 Ultra myself. I have never experienced any issues. All apps keep running.
But you have listed Samsung as worst.

I can confirm that I get the most mails from users using Xiaomi and Huawei complaining that the app stops working in the background. In that regard your library is right.
But Samsung needs to be moved down from the top.
Users thinking that they have a better device than getting Samsung, but it's not true..

Thank you,
Tim K.

**Comment by petrnalevka** [comment]
Hello Tim, big thanks for your feedback on this. I agree that Samsung is not the worst in terms of non-standard killing practices, but because of its market share it is definitely one of the biggest headaches for devs IMHO. It seems everything works as expected unless users let the system move your app to the list of sleeping apps when background processing basically stops for the app. Maybe you as an experienced user can see when this is happening, can properly decipher the notification which precedes this but most users out there won't. For instance for me as an alarm clock app maintainer this is major problem. When users do not use the app fro few days their alarm won't trigger at the next day and users have abosolutely no idea why. This "extension" to the AOSP adaptive battery mechanism is IMHO really harmful. Maybe you could comment on this mechanism as I'm not familiar with all details on how is this implemented with latest Samsung versions, but we are still seeing this feedback. On the other hand I agree that Xiaomi, Huawei and 1+ are more problematic, but luckily with a much smaller market share which makes them less of a problem for devs..

**Comment by petrnalevka** [comment]
Hello Tim, big thanks for your feedback on this. I agree that Samsung is
not the worst in terms of non-standard killing practices, but because of
its market share it is definitely one of the biggest headaches for devs
IMHO. It seems everything works as expected unless users let the system
move your app to the list of sleeping apps when background processing
basically stops for the app. Maybe you as an experienced user can see when
this is happening, can properly decipher the notification which precedes
this but most users out there won't. For instance for me as an alarm clock
app maintainer this is major problem. When users do not use the app fro few
days their alarm won't trigger at the next day and users have abosolutely
no idea why. This "extension" to the AOSP adaptive battery mechanism is
IMHO really harmful. Maybe you could comment on this mechanism as I'm not
familiar with all details on how is this implemented with latest Samsung
versions, but we are still seeing this feedback. On the other hand I agree
that Xiaomi, Huawei and 1+ are more problematic, but luckily with a much
smaller market share which makes them less of a problem for devs..

Thank you.

Best regards,

Petr


---
P.S. Sorry for my brevity or longer response time. I'm handling hundreds of
emails daily and I try hard to respond to most of them in the best way I
can. Just let me know if you need more details, please.

**Comment by rl885** [comment]
Hello Tim, on our support, Samsung still takes twice as many reports as the other vendors (the last year). This might be because they have a bigger market share, but I cannot agree with "Samsung is the best" in terms of background restrictions. 
The restrictions are there and enabled by default. On most phones, they are hidden all over the place and users have trouble finding them in the layout.
Moving Samsung from the first position is fair, but we are still too far from marking Samsung as the best vendor for background management.

**Comment by T1mmyK0** [comment]
Yes, Samsung has this sleep mode for apps not being used for a while. But I think this is less of a problem. Because you need to not open an app for multiple months for it to get stopped. And once you wonder why it isn't working anymore, you will open it manually and it will work again. Also, app developers can easily warn users. On Samsung devices changing this setting manually definitely works.
On Xiaomi and Huawei devices, users need to change multiple settings and they regularly tell me that changing these settings doesn't really do anything.
Also, on these devices an app will stop working from the first day on. Users will always think it's the apps fault. If it gets stopped after a few months of not being used, most users will notice that it's not the app, but their device settings.
Same with messenger apps which you open regularly anyways. I have a friend using Huawei. She doesn't get Signal notification. We contact her, but she doesn't even notice. On Samsung you basically never experience issues with a messenger.

Where are your website visitors from?
I think it depends heavily on the countries. In Europe most will probably use Samsung.
But I have a lot of users from poorer countries like Brazil, Mexico, Indonesia and India. They are using Xiaomi and Huawei a lot which is a huge problem.
I think Xiaomi comes right after Samsung in terms of market share in those countries.

**Comment by rl885** [comment]
If you have alarms, calendars, or other time-sensitive events, you cannot rely on remembering to open the app actively every 3 days so the system won't suspend it in the background. 
If there is a need to warn users, that the system may kill the app sometime in the future without warning, then the phone deserves a bad rating. The users should not be forced to dig deep into the settings to make sure the system won't decide to kill their apps.
Why does not the system warn the users "This app does not seem to be used, do you want to suspend it?"... 

Samsung also reverts the settings, that were already toggled by the users - this has been reported many times.

If the app stops working, the users will ALWAYS blame the app. Half of the issues we are dealing with start with "after using the app for x years, it suddenly stopped working" (typically a 1* review).

I moved Samsung to the 4th position, but until Samsung stops putting so many restrictions under different names and in different places in the settings, we don't want to give it a better rating.

**Comment by petrnalevka** [comment]
Hello Tim, not sure if this period is a months on S22, but definitely on some Samsung devices this period is as short as 3 days.. So imagine an alarm clock app not being used on the weekend and one additional day and whoops you won't get woken up next day. From our personal experience at least half of app killing issues reported on our support are from Samsung. If you want some more exact measures, we have results from our benchmarking app DontKillMyApp and half of scores under 10% are from Samsung devices. I do not expect users would leave the benchmark up for 1 month before they run the bechmark. Anyway, me are changing the order of the top app killers slightly we this is what we have felt for a longer tiem as well that the current order is not completely right.. For various reasons it is tiem for Xiaomi to get 5/5 as in recent updates they are even killing accessibility servces! And Samsung maybe really does not deserve to be at spot one, even from our data it seems it causes most trouble to devs..

**Comment by T1mmyK0** [comment]
Thank you, that's totally fine. Samsung still has issues. I get a little annoyed by the unused app stopping as well. But definitely Xiaomi and Huawei needs to be ranked worse. I can explain this further if you like.
Xiaomi and Huawei even have this stupid Autostart settings where apps are not allowed to start after a device restart. This is bad for any app that needs to do something for the user. And then users also tell me the app stops after midnight. Biggest problem is that Xiaomi and Huawei don't even start the app correctly in the wakeup phase again. Samsung does that.

Let me also explain what app I have. Its an auto reply app which replies to notifications. So it uses the notification listener service. My app technically cannot start its service. It can only let it start by the system. It's directly connected to the notification access permission. So users open the setting and see that my app is enabled there. But the service isn't running in the background. Even with my app open in the foreground. Xiaomi and Huawei just don't start it again after stopping it. Users need to revoke the permission and grant it again to make it work. And with their bad Autostart setting not even a device restart helps. Their software is so bad that they cannot even differ between normal app services and services that are connected to triggers like notifications. My app service actually isn't running permanently. It only needs to run for a very short time after a notification is received. The same problem with accessibility services as you already told. Killing such services just doesn't make any sense at all :/
